raft.go

```diff
@@ -36,38 +36,72 @@ func (a *Autopilot) NumVoters() (int, error) {
 // its all done will trigger autopilot to remove dead servers if there
 // are any. Servers added by this method will start in a non-voting
 // state and later on autopilot will promote them to voting status
-// if desired by the configured promoter.
+// if desired by the configured promoter. If too many removals would
+// be required that would cause leadership loss then an error is returned
+// instead of performing any Raft configuration changes.
 func (a *Autopilot) AddServer(s *Server) error {
 	cfg, err := a.getRaftConfiguration()
 	if err != nil {
 		a.logger.Error("failed to get raft configuration", "error", err)
 		return err
 	}
 
+	var existingVoter bool
+	var voterRemovals []raft.ServerID
+	var nonVoterRemovals []raft.ServerID
+	var numVoters int
 	for _, server := range cfg.Servers {
+		if server.Suffrage == raft.Voter {
+			numVoters++
+		}
+
 		if server.Address == s.Address && server.ID == s.ID {
 			// nothing to be done as the addr and ID both already match
 			return nil
-		} else if server.Address == s.Address || server.ID == s.ID {
-			// we know that we have a partial match  so we should remove then re-add the server.
-			if err := a.removeServer(server.ID); err != nil {
-				if server.ID == s.ID {
-					return fmt.Errorf("error removing server with duplicate ID %q: %w", server.ID, err)
-				} else {
-					return fmt.Errorf("error removing server with duplicate address %q: %w", server.Address, err)
-				}
+		} else if server.ID == s.ID {
+			// special case for address updates only. In this case we should be
+			// able to update the configuration without have to first remove the server
+			if server.Suffrage == raft.Voter || server.Suffrage == raft.Staging {
+				existingVoter = true
 			}
-
-			if server.ID == s.ID {
-				a.logger.Info("removed server with duplicate ID", "id", server.ID)
+		} else if server.Address == s.Address {
```

Comment:
Here we are going to build up the list of servers that would require removal due to conflicts with the node being added. The max here is two removals as we could at most have a conflicts for the id and address. We don't want to immediately  remove them because further down we want to ensure the removals would be safe.

Reply:
Run that by me again? Since this conditional is just checking if the address conflicts with the same ID, how could we have 2 removals? That would mean multiple servers already had the same address.

Reply:
Actually we wont have two removals. We will have a removal + an update. So really there can only ever be one removal but as we are looping through things it seemed best to keep a list process it later on.